<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Вводные-данные:" data-toc-modified-id="Вводные-данные:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Вводные данные:</a></span></li><li><span><a href="#Цель-исследования:" data-toc-modified-id="Цель-исследования:-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Цель исследования:</a></span></li><li><span><a href="#План-исследования:" data-toc-modified-id="План-исследования:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>План исследования:</a></span></li><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года;" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года;-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года;</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку;</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

## Вводные данные:

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать. Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. База данных приложения содержит информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

## Цель исследования:

Задача - проанализировать базу данных приложения, для тех кто любит читать для дальнейшей формулировки ценностного предложения для нового продукта.

## План исследования:

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [2]:
# выведем таблицу с книгами
books = '''SELECT * FROM books LIMIT 5'''

pd.io.sql.read_sql(sql=text(books), con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
# выведем таблицу с авторами книг
authors = '''SELECT * FROM authors LIMIT 5'''

pd.io.sql.read_sql(sql=text(authors), con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
# выведем таблицу с пользовательскими оценками книг:
ratings = '''SELECT * FROM ratings LIMIT 5'''

pd.io.sql.read_sql(sql=text(ratings), con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [5]:
# выведем таблицу с пользовательскими обзорами на книги
reviews = '''SELECT * FROM reviews LIMIT 5'''

pd.io.sql.read_sql(sql=text(reviews), con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [6]:
# выведем таблицу с издательствами
publishers = '''SELECT * FROM publishers LIMIT 5'''

pd.io.sql.read_sql(sql=text(publishers), con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

## Посчитайте, сколько книг вышло после 1 января 2000 года;


In [7]:
book_count = '''SELECT COUNT(book_id) 
                FROM books
                WHERE publication_date > '2000-01-01'
                '''

pd.io.sql.read_sql(sql=text(book_count), con = engine)

count
0    819

После 1 января 2000 года вышло 819 книг.

## Для каждой книги посчитайте количество обзоров и среднюю оценку;


In [8]:
book_count = '''SELECT  DISTINCT title,
                b.book_id,
                COUNT(DISTINCT r.review_id) as review_count,
                AVG(ra.rating) as avg_rating
                FROM books as b
                LEFT JOIN ratings AS ra ON ra.book_id=b.book_id
                LEFT JOIN reviews AS r ON b.book_id=r.book_id
                GROUP BY b.book_id
                ORDER BY 3 DESC
                '''

pd.io.sql.read_sql(sql=text(book_count), con = engine)

title  book_id  review_count  \
0                              Twilight (Twilight  #1)      948             7   
1                                      Eat  Pray  Love      207             6   
2    Harry Potter and the Chamber of Secrets (Harry...      299             6   
3    Harry Potter and the Prisoner of Azkaban (Harr...      302             6   
4                            Outlander (Outlander  #1)      497             6   
..                                                 ...      ...           ...   
995  Disney's Beauty and the Beast (A Little Golden...      191             0   
996                          Essential Tales and Poems      221             0   
997                               Leonardo's Notebooks      387             0   
998   The Cat in the Hat and Other Dr. Seuss Favorites      672             0   
999                            The Natural Way to Draw      808             0   

     avg_rating  
0      3.662500  
1      3.395833  
2      4.287500  
3      4.414634  
4      4.125000  
..          ...  
995    4.000000  
996    4.000000  
997    4.000000  
998    5.000000  
999    3.000000  

[1000 rows x 4 columns]

In [9]:
book_count = '''SELECT  DISTINCT title,
                b.book_id,
                COUNT(DISTINCT r.review_id) as review_count,
                AVG(ra.rating) as avg_rating
                FROM books as b
                LEFT JOIN ratings AS ra ON ra.book_id=b.book_id
                LEFT JOIN reviews AS r ON b.book_id=r.book_id
                GROUP BY b.book_id
                ORDER BY 4 ASC
                '''

pd.io.sql.read_sql(sql=text(book_count), con = engine)['review_count'].sum()

2793

Всего 1000 книг, средняя оценка варьируется от 5.00 до 1.50. Максимальное количество обзоров - 7, минимальное - 0.

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [10]:

publisher_top = '''SELECT p.publisher,
                COUNT(b.book_id)
                FROM publishers AS p
                JOIN books AS b ON b.publisher_id=p.publisher_id
                WHERE b.num_pages>50
                GROUP BY 1
                ORDER BY COUNT(b.book_id) DESC
                LIMIT 1
                '''

pd.io.sql.read_sql(sql=text(publisher_top), con = engine)


publisher  count
0  Penguin Books     42

Наибольшее число книг толще 50 страниц выпустило издательство  Penguin Books -  42 книги.

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [11]:

author_top = '''SELECT b.author_id,
                   a.author,
                   AVG(r.rating)
                   FROM books AS b
                   JOIN ratings AS r ON r.book_id=b.book_id
                   JOIN authors AS A ON a.author_id=b.author_id
                   GROUP BY b.author_id,a.author
                   HAVING COUNT(r.rating) > 50
                   ORDER BY AVG(r.rating) DESC
                   LIMIT 1
                   
                '''

pd.io.sql.read_sql(sql=text(author_top), con = engine)


author_id                      author       avg
0        236  J.K. Rowling/Mary GrandPré  4.288462

Автор с самой высокой средней оценкой книг среди книг, получивших 50 и более оценок - Джоан Роулинг и иллюстратор Мари Гранпэ. Гарри Поттер пользуется популярностью.

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [12]:

publisher_top = '''
                   WITH first AS ( SELECT v.username,
                                   COUNT(v.review_id) AS reviews_count_per_user
                                   FROM reviews AS v
                                   WHERE v.username IN (SELECT r.username
                                                       FROM ratings AS r
                                                       GROUP BY r.username
                                                       HAVING COUNT(r.username)>50 )
                                   GROUP BY   v.username)
                   SELECT  AVG (reviews_count_per_user)
                   FROM first
                      
                '''

pd.io.sql.read_sql(sql=text(publisher_top), con = engine)

avg
0  24.333333

Среднее количество обзоров от пользователей поставивших более 50 оценок - 24.

## Общий вывод

- После 1 января 2000 года вышло 819 книг.
- Всего 1000 книг, средняя оценка варьируется от 5.00 до 1.50. Максимальное количество обзоров - 1120, минимальное - 0.
- Наибольшее число книг толще 50 страниц выпустило издательство  Penguin Books -  42 книги.
- Автор с самой высокой средней оценкой книг среди книг, получивших 50 и более оценок - Джоан Роулинг и иллюстратор Мари Гранпэ. 
- Среднее количество обзоров от пользователей поставивших более 50 оценок - 24.